File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        N = len(nums)
        nums.sort()
        output = []
        i = 0
        for i in range(N-2):
            if i != 0 and nums[i] == nums[i-1]:
                continue

            target = - nums[i]
            l, r = i+1, N-1
            while l < r:
                total = nums[l]+nums[r]
                if total == target:
                    output.append([nums[i], nums[l], nums[r]])
                    curL, curR = nums[l], nums[r]
                    while nums[l] == curL and l < r:
                        l += 1
                    while nums[r] == curR and l < r:
                        r -= 1
                elif total < target:
                    l += 1
                else:
                    r -= 1



        return output


        

File: 15-3sum/3sum.dart
class Solution {
  List<List<int>> threeSum(List<int> nums) {
    nums.sort();
    List<List<int>> ans = [];

    //Ana for loop
    for (int i = 0; i < nums.length-2; i++){
        // Ayni sayiyi tekrar etmemek icin
        if (i > 0 && nums[i] == nums[i-1]) continue;

        // Alt dizide 2 sum ariyoruz
        int l = i+1;
        int r = nums.length-1;
        int target = -nums[i];

        while (l < r){
            int sum = nums[l] + nums[r];
            if(sum < target){
                l++;
            }else if(sum > target){
                r--;
            }else{
                //Cozum bulundu
                ans.add([nums[i],nums[l],nums[r]]);
                int curL = nums[l];
                int curR = nums[r];

                while (l < r && nums[l] == curL) l++;
                while (l < r && nums[r] == curR) r--;
            }
        }
    }
    return ans;
    
  }
}